### PR TITLE
fix: implement native_iceberg_drop

### LINKED ISSUE #285: Attempts to "clean up" s3 in case of Iceberg tables, which is redundant
More new behaviour since 1.3.5

For CTAS dbt-athena attempts to delete objects at location in the case of iceberg table. It did not do that before, so this is another breaking change thus requiring new S3 permissions (listing and deleting) if you're using iceberg (which I am, and I don't have the necessary permissions).

This activity seems to be redundant given [AWS docs](https://docs.aws.amazon.com/athena/latest/ug/querying-iceberg-managing-tables.html)

> Because Iceberg tables are considered managed tables in Athena, dropping an Iceberg table also removes all the data in the table.

Problematic line appears to be [here](https://github.com/dbt-athena/dbt-athena/blob/af91c89ea669786e3026ec5a00becb9612733776/dbt/include/athena/macros/materializations/models/table/create_table_as.sql#L51). I'll PR a fix to remove this call in case of iceberg.

### LINKED ISSUE #291: Document minimum necessary permissions in testable form
See https://github.com/dbt-athena/dbt-athena/issues/66#issuecomment-1538056731

As a dbt-athena contributor
I want an IAM role defined in the source that describes the minimal permissions needed to execute dbt-athena
So that I can test against those permissions to detect and avoid breaking changes

As a dbt-athena consumer
I want an IAM role defined in the source that describes the minimal permissions needed to execute dbt-athena
So that I understand what permissions I need to make available to principals executing dbt-athena

Acceptance Criteria:
- IAM role exists describing minimal permissions
- Documentation references IAM role for contributors and consumers
- Documentation describes how to test with the role
- OPTIONAL make provides option to deploy the role to target AWS account
- OPTIONAL functional tests automatically assume the role

### PR DESCRIPTION
**Pull request by brabster**
… dev and testing

### Description

- resolve #285 
- resolve #291 

Remove need for direct access to S3 & Glue to drop an Iceberg table. I figured that there needs to be a way for me and others to test the solution works with the restricted permissions so I have the start of that here.

## Checklist
- [/] You followed [contributing section](https://github.com/dbt-athena/dbt-athena#contributing)
- [/] You kept your Pull Request small and focused on a single feature or bug fix.
- [ ] You added unit testing when necessary
- [ ] You added functional testing when necessary

### CODE REVIEW

File: tests/restrictive_role_cf.yaml [hunk 1]
```diff
@@ -0,0 +1,80 @@
+
```
Comment (nicor88) [inline tests/restrictive_role_cf.yaml:1]: before you spend more effort on this, could you give a general idea on what you plan to implement? 
ideally all the necessary infra to run integration test is not implemented in dbt-athena, but in https://github.com/dbt-athena/dbt-athena-infra
then in dbt-athena CI we will assume role deployed somewhere else, and we could consider to also cover this use case ;)

Comment (nicor88) [reply tests/restrictive_role_cf.yaml:1]: I think that supporting `native_drop` for iceberg is the way to got.
At some point, Athena will improve/enhance the limitations on removing all the s3 objects on bigger datasets, and if that happens having this feature backed in will be beneficial for use cases like yours, but at the same time could mean a change of a flag for users that want to benefit from it.

**Comment by brabster** [comment]
:thinking: when I committed the restrictive role contained in the CloudFormation template for convenience) my thinking was to document a specific set of permissions in a role that can be used by anyone (or automation) for testing that the base functionality works without direct access to restricted APIs - obvs. could loosen the restrictions eg. to allow direct access to certain Glue APis.

**Comment by nicor88** [comment]
> Maybe it's a separate thread/issue, but having a way to explicitly test that basic dbt functionality works under a more restrictive set of permissions than the [one currently used](https://github.com/dbt-athena/dbt-athena-infra/blob/0d5b6a060e60a6d255fcc5f946906490b0c6b400/lib/stacks/github-stack.ts#L119) seems like the only way to really say that dbt-athena can be used under those permissions?
that's indeed the right thought. 
What do we need to achieve this in a more scalable way:
* a IAM role that can be assumed by github action with resticted permissions - I would add it here: https://github.com/dbt-athena/dbt-athena-infra (CDK based setup) - see [here](https://github.com/dbt-athena/dbt-athena-infra/blob/main/lib/stacks/github-stack.ts#L59-L64) 
* another section in [functional-tests](https://github.com/dbt-athena/dbt-athena/blob/main/.github/workflows/functional-tests.yml) that assume the role with restricted mode and we the flag native_drop=True


IMHO, in this specific case you are not really blocked, as you can prepare the feature without all the testing setup - to see if it works we can check if drops query hits athena, and you can turn debug mode on to see if the s3 cleanup path is called.

**Comment by brabster** [comment]
It might make sense to combine this and the native seed thing - they are a bit coupled together which makes me a bit unsure whether it's good as a whole

Evidence
```
[0m21:37:10.808696 [debug] [Thread-1  ]: Began executing node seed.ephemeral.base
[0m21:37:10.881750 [debug] [Thread-1  ]: Opening a new connection, currently in state closed
[0m21:37:11.224528 [debug] [Thread-1  ]: Dropping relation via SQL only
[0m21:37:11.226145 [debug] [Thread-1  ]: Using athena connection "seed.ephemeral.base"
[0m21:37:11.227143 [debug] [Thread-1  ]: On seed.ephemeral.base: drop table if exists `awsdatacatalog`.`test16843558268112747602_test_basic_iceberg_native_drop`.`base__dbt_tmp`
    
[0m21:37:12.694506 [debug] [Thread-1  ]: SQL status: OK -1 in 1.0 seconds
[0m21:37:12.714119 [debug] [Thread-1  ]: Using athena connection "seed.ephemeral.base"
[0m21:37:12.715985 [debug] [Thread-1  ]: On seed.ephemeral.base: 
    create external table `awsdatacatalog`.`test16843558268112747602_test_basic_iceberg_native_drop`.`base__dbt_tmp` (id string, name string, some_date string)
    row format serde 'org.apache.hadoop.hive.serde2.OpenCSVSerde'
    location 's3://843328850426-athena-results/staging/tables/test16843558268112747602_test_basic_iceberg_native_drop/base__dbt_tmp/afd3b34c-0d4c-4581-82d4-da916d517f73'
    tblproperties (
      'skip.header.line.count'='1'
    )
  
[0m21:37:14.046299 [debug] [Thread-1  ]: SQL status: OK -1 in 1.0 seconds
[0m21:37:14.362476 [debug] [Thread-1  ]: Config native_drop enabled, skipping direct S3 delete
[0m21:37:14.363567 [debug] [Thread-1  ]: Using athena connection "seed.ephemeral.base"
```

File: dbt/include/athena/macros/adapters/relation.sql [hunk 2]
```diff
@@ -1,8 +1,30 @@
 {% macro athena__drop_relation(relation) -%}
+  {%- set native_drop = config.get('native_drop', default='False') | as_bool -%}
+
+  {%- if native_drop -%}
+    {%- do drop_relation_sql(relation) -%}
```
Comment (nicor88) [inline dbt/include/athena/macros/adapters/relation.sql:5]: Native drops works only for iceberg, this means if a user use the native drop globally and it uses hive tables, data in s3 won't be clean.
We should add another condition in case of native drop that check the relation type, if we deal with iceberg tables or views we execute only drop, if the relation is a table we want to clean up also s3.

Comment (brabster) [reply dbt/include/athena/macros/adapters/relation.sql:6]: Addressed

**Comment by brabster** [comment]
Test run output, unit_tests all passing too
```bashsession
pytest -n auto tests/functional
====================================================================================== test session starts ======================================================================================
platform linux -- Python 3.8.13, pytest-7.3.1, pluggy-1.0.0
rootdir: /home/paul/projects/brabster/dbt-athena
configfile: pytest.ini
plugins: cov-4.0.0, dotenv-0.5.2, xdist-3.3.0
8 workers [57 items]    
ssssss....ssss.ss.s.....ss........s......................                                                                                                                                 [100%]
======================================================================================= warnings summary ========================================================================================
tests/functional/adapter/test_basic_hive.py: 31 warnings
tests/functional/adapter/test_basic_hive_native_drop.py: 15 warnings
tests/functional/adapter/test_basic_iceberg.py: 15 warnings
tests/functional/adapter/test_basic_iceberg_native_drop.py: 6 warnings
tests/functional/adapter/utils/test_utils.py: 60 warnings
tests/functional/adapter/test_snapshot.py: 36 warnings
  /home/paul/projects/brabster/dbt-athena/venv/lib/python3.8/site-packages/jinja2/lexer.py:650: DeprecationWarning: invalid escape sequence '\('

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=================================================================== 41 passed, 16 skipped, 163 warnings in 593.48s (0:09:53) ====================================================================
```

File: tests/restrictive_role_cf.yaml (hunk 1 above)
Comment (brabster) [reply tests/restrictive_role_cf.yaml:1]: Adopted native_drop suggestion

File: dbt/include/athena/macros/adapters/relation.sql [hunk 3]
```diff
@@ -1,8 +1,32 @@
 {% macro athena__drop_relation(relation) -%}
+  {%- set native_drop = config.get('drop', default='False') | as_bool -%}
+  {%- set natively_droppable = config.get('table_type', default='hive') == 'iceberg' or relation.type == 'view' -%}
```
Comment (nicor88) [inline dbt/include/athena/macros/adapters/relation.sql:3]: This unfortunately doesn't work in all cases. The source of truth is not what we have in the config but what we have in the glue catalog. We had this implementation back then, and it leads to some issues.
Have a look at this refactor https://github.com/dbt-athena/dbt-athena/pull/249

Specifically, we called
`{% set rel_type = adapter.get_table_type(relation.schema, relation.table) %}` to fetch table type via an adapter method
(the method doesn't exist anymore), we need to put it back (cc @Jrmyy) and use the [this](https://github.com/dbt-athena/dbt-athena/blob/main/dbt/adapters/athena/relation.py#L92-L104) helper to get the right enum type, then we can check if we can drop natively or not.

There are edge cases where this is needed, for example if have a relation that is an hive table, and then the config is changed to iceberg table, we want the adapter drop the table and then clean up s3, but based in this implementation config is used, there the s3 clean will be used, leading to some dirty s3 path - same with views.
Long story short, the issue is that we need to rely on what "we" have (data catalog), not what we plan to have after run (config).

**Review by nicor88 — CHANGES_REQUESTED** [review]
Please see my comment.

File: dbt/include/athena/macros/adapters/relation.sql (hunk 3 above)
Comment (brabster) [reply dbt/include/athena/macros/adapters/relation.sql]: Applied changes:
- picked up old impl method
- picked up old tests
- replaced config get with table type get
- confirmed as best I can in logs, tests all pass

File: tests/restrictive_role_cf.yaml (hunk 1 above)
Comment (brabster) [reply tests/restrictive_role_cf.yaml]: Moved role defs to infra repo

File: dbt/include/athena/macros/materializations/models/table/create_table_as.sql [hunk 4]
```diff
@@ -20,6 +20,7 @@
                                                  s3_data_naming,
                                                  external_location,
                                                  temporary) -%}
+  {%- set native_drop = config.get('native_drop', default='False') | as_bool -%}
```
Comment (svdimchenko) [inline dbt/include/athena/macros/materializations/models/table/create_table_as.sql:23]:
```suggestion
  {%- set native_drop = config.get('native_drop', 'False') | as_bool -%}
```
due to https://stackoverflow.com/questions/24463202/typeerror-get-takes-no-keyword-arguments

File: dbt/adapters/athena/impl.py [hunk 5]
```diff
@@ -786,3 +787,21 @@ def _generate_snapshot_migration_sql(self, relation: AthenaRelation, table_colum
                 drop_staging_sql.strip(),
             ]
         )
+
+    @available
+    def get_table_type(self, db_name, table_name) -> TableType:
```
Comment (svdimchenko) [inline dbt/adapters/athena/impl.py:792]: we already have this method in `relation` module. Seems that we don't need such code duplication. Does it make sense to move this method to `AthenaRelation` class if you need it to be used inside macro ? @Jrmyy WDYT ?

Comment (Jrmyy) [reply dbt/adapters/athena/impl.py:792]: Hey,
Multiple things here:
1. We have a standard now that required, whenever possible, to pass the relation object to the function instead of db_name and table_name
2. We already do something similar inside `get_glue_table_location`. I would suggest to find a way to reuse part of this code, or at least to prevent from duplication

File: dbt/include/athena/macros/adapters/relation.sql [hunk 6]
```diff
@@ -1,8 +1,34 @@
...
 {% endmacro %}
 
+{% macro drop_relation_sql(relation) -%}
+
+  {%- do log('Dropping relation via SQL only') -%}
+  {% call statement('drop_relation', auto_begin=False) -%}
+    {%- if relation.type == 'view' -%}
+      drop {{ relation.type }} if exists {{ relation.render() }}
+    {%- else -%}
+      drop {{ relation.type }} if exists {{ relation.render_hive() }}
```
Comment (svdimchenko) [inline dbt/include/athena/macros/adapters/relation.sql:27]: how do you handle `DDL execution timed out` issue here ?

Comment (brabster) [reply dbt/include/athena/macros/adapters/relation.sql:27]: I have raised an issue [here](https://github.com/dbt-athena/dbt-athena/issues/312) to look at that separately. I can add something in the docs around this flag to call out the potential issue. I'm not actually seeing timeouts generally in my real dbt project (maybe 30-40 Iceberg tables being dropped + recreated several times a day) but it seems clear it's an issue in AWS' implementation of Athena, not a dbt-athena issue.

File: dbt/adapters/athena/impl.py (hunk 5 above)
Comment (brabster) [reply dbt/adapters/athena/impl.py]: Refactoring completed, see latest. Unit/func tests all passing for me

File: dbt/adapters/athena/impl.py [hunk 7]
```diff
@@ -186,10 +187,9 @@ def generate_s3_location(
         return mapping[s3_data_naming]
 
     @available
-    def get_glue_table_location(self, relation: AthenaRelation) -> Optional[str]:
+    def get_glue_table(self, relation: AthenaRelation) -> Optional[str]:
```
Comment (svdimchenko) [inline dbt/adapters/athena/impl.py:190]:
```suggestion
    def get_glue_table(self, relation: AthenaRelation) -> Optional[GetTableResponseTypeDef]:
```
see `mypy_boto3_glue.type_defs`

**Review by svdimchenko — APPROVED** [review]
@brabster the code LGTM, great job done with tests implementation. @nicor88 @Jrmyy could you take a look at this PR as well ?

**Comment by svdimchenko** [comment]
please, fix the CI with required type import `GetTableResponseTypeDef`

**Review by svdimchenko — CHANGES_REQUESTED** [review]
please, fix the CI with required type import `GetTableResponseTypeDef`

**Comment by brabster** [comment]
Not sure how to close off the 2x requested changes - they have been implemented

**Review by svdimchenko — APPROVED** [review]

**Review by Jrmyy — APPROVED** [review]

**Comment by brabster** [comment]
I've been trying out these changes in my real project and looking at the logs I've got a suspicion that native_drop might be taking effect when it shouldn't... super difficult to be confident without tests and it's unclear how to write/augment tests that ensure the correct behaviour in on/off cases.

I did wonder about asserting the logs do or do not contain specific text (this is how I'm checking by hand) but I'm not sure how to assert against the caplog for the existing test cases that are extending dbt base tests, any ideas?

**Comment by brabster** [comment]
There was a bug, the way I was trying to get the value of 'native_drop' as a boolean was always true and that meant that any iceberg table would get the native drop behaviour regardless of the flag.

```
{%- set native_drop = config.get('native_drop', default='False') | as_bool -%}
```

Swapped to

```
{%- set native_drop = config.get('native_drop', 'False') == 'True' -%}
```

Which seems to work correctly based on log inspection. Also removed a few messy log messages that I'd left in there.

**Comment by brabster** [comment]
@svdimchenko my adjustments are based on several trials of different ways of expressing a boolean such that it worked in the iceberg test cases and in my actual model dbt config. I was really struggling to work out what the heck type was actually coming out in the Jinja template until I found that sqlserver example and [this SO post](https://stackoverflow.com/questions/8433450/why-doesnt-my-condition-logic-work-as-expected-in-jinja2-cherrypy) which refers to [Jinja docs saying](https://jinja.palletsprojects.com/en/3.1.x/templates/) and suggests that the bool types are lowercased in Jinja:

> The special constants true, false, and none are indeed lowercase. Because that caused confusion in the past, (True used to expand to an undefined variable that was considered false), all three can now also be written in title case (True, False, and None). However, for consistency, (all Jinja identifiers are lowercase) you should use the lowercase versions.

So i'm confused now - especially as `config.get('foo', default='bar')` form is used everywhere else in the templates? Maybe I'm misunderstanding?

**Review by svdimchenko — APPROVED** [review]

File: README.md [hunk 8]
```diff
@@ -134,6 +134,7 @@ _Additional information_
 * `field_delimiter` (`default=none`)
   * Custom field delimiter, for when format is set to `TEXTFILE`
 * `table_properties`: table properties to add to the table, valid for Iceberg only
++ `native_drop`: Relation drop operations will be performed with SQL, not direct Glue API calls. No S3 calls will be made to manage data in S3. Data in S3 will only be cleared up for Iceberg tables [see AWS docs](https://docs.aws.amazon.com/athena/latest/ug/querying-iceberg-managing-tables.html). Note that Iceberg DROP TABLE operations may timeout if they take longer than 60 seconds.
```
Comment (Jrmyy) [inline README.md:137]: Can you also make a PR in our [documentation](https://github.com/dbt-athena/dbt-athena.github.io/) repo if it is not done already ? 🙂

**Review by Jrmyy — DISMISSED** [review]

File: README.md (hunk 8 above)
Comment (brabster) [reply README.md:137]: Sure no problem... do you have any idea when this will merge to main?

Comment (Jrmyy) [reply README.md:137]: I think we can merge this PR now

Comment (svdimchenko) [reply README.md:138]: @Jrmyy I've resolved conflict and we can merge it now
